039-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,264 @@
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care of running a rebalancing operation automatically if requested.
+
+Taking into account the `StrimziPodSets` usage (instead of the `StatefulSets`), which is the default since the [0.30.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.30.0) release, a more general use case will be about adding new brokers and/or removing existing ones, not talking just about scaling up and down anymore.
+
+This proposal takes into account this future plan by referring to more general operations like add and remove brokers instead of scaling up and down.
+
+## Proposal
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the configuration for the auto-rebalancing but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance` one.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, of adding the configuration to the `Kafka` custom resource, this will avoid polluting the `Kafka` custom resource and increasing its size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+spec:
+  # NOTE: mode and brokers fields are not available as in the KafkaRebalance,
+  #       they are automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) the same way they already do today on a `KafkaRebalance` custom resource.
+This allows the user to not rely on the default Cruise Control goals configuration which would be common for any rebalance without overriding goals.
+They could define different goals based on the status of the Kafka cluster or skip hard goals in some scenarios for example.
+As already mentioned, specifying goals but more in general customizing the optimization proposal request is already possible in the current usage of a `KafkaRebalance` custom resource so it should be allowed for the auto-rebalancing as well.
+In the end, the cluster operator will use that configuration from the `KafkaRebalanceTemplate` to fill the corresponding `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
```

Comment:
What will happen to the auto-generated `KafkaRebalance` once the rebalance is complete? Will it be left so the user can see what happened, or deleted so it isn't cluttering up the cluster?